packages/rich-text/src/to-dom.js

```diff
@@ -0,0 +1,230 @@
+/**
+ * Internal dependencies
+ */
+
+import { toTree } from './to-tree';
+
+/**
+ * Browser dependencies
+ */
+
+const { TEXT_NODE, ELEMENT_NODE } = window.Node;
+
+/**
+ * Creates a path as an array of indices from the given root node to the given
+ * node.
+ *
+ * @param {Node}        node     Node to find the path of.
+ * @param {HTMLElement} rootNode Root node to find the path from.
+ * @param {Array}       path     Initial path to build on.
+ *
+ * @return {Array} The path from the root node to the node.
+ */
+function createPathToNode( node, rootNode, path ) {
+	const parentNode = node.parentNode;
+	let i = 0;
+
+	while ( ( node = node.previousSibling ) ) {
+		i++;
+	}
+
+	path = [ i, ...path ];
+
+	if ( parentNode !== rootNode ) {
+		path = createPathToNode( parentNode, rootNode, path );
+	}
+
+	return path;
+}
+
+/**
+ * Gets a node given a path (array of indices) from the given node.
+ *
+ * @param {HTMLElement} node Root node to find the wanted node in.
+ * @param {Array}       path Path (indices) to the wanted node.
+ *
+ * @return {Object} Object with the found node and the remaining offset (if any).
+ */
+function getNodeByPath( node, path ) {
+	path = [ ...path ];
+
+	while ( node && path.length > 1 ) {
+		node = node.childNodes[ path.shift() ];
+	}
+
+	return {
+		node,
+		offset: path[ 0 ],
+	};
+}
+
+function createEmpty( type ) {
+	const { body } = document.implementation.createHTMLDocument( '' );
```

Comment:
This has frequently turned up in performance profiling as an expensive operation, which on reflection is understandable given that it's creating an entirely new document. I think we should hold on to a constant reference of a document body and reset it each time we call `createEmpty`.

Benchmark: [Creating a "fresh" body using `innerHTML` to reset is **646x** faster than via `document.implementation.createHTMLDocument`](https://jsbin.com/pekokug/4/edit?html,js,console,output)

Reply:
Yes, I saw the other PR. We can do the same here.